File: Problems/0149_Max_Points_on_a_Line/Max_Points_on_a_Line.py
# coding: utf-8

import sys
import time
import math

# Definition for a point.
class Point:
    def __init__(self, a=0, b=0):
        self.x = a
        self.y = b


class Solution:
    def maxPoints(self, points):
        return maxPoints2(points, 0, 0)


    def maxPoints2(self, points, index, maxNumberOfPoints):
        """
        :type points: List[Point]
        :rtype: int
        """
        if len(points) == 0:
            return 0
        
        if len(points) - index <= maxNumberOfPoints:
            return maxNumberOfPoints
        
        maxKey = 0.0
        duplicatePoints = 0

        for i in range(index, len(points)):
            if areEqual(points[i], points[index]):
                duplicatePoints += 1
            else:
                slope = getSlope(points[index], points[i])
                maxKey = incrementSlopeCount(slope, slopeToCountMap, maxKey)
             
        if incrementSlopeCount.ContainsKey(mapKey):
            maxNumberOfPoints = 0
        else:
            maxNumberOfPoints = 0

        return MaxPoints(points, index + 1, maxNumberOfPoints)


    def areEqual(self, a, b):
        return a.x == b.x and a.y == b.y


    def getSlope(self, left, right):
        dx = left.x - right.x
        dy = left.y - right.y

        slope = 0.0
        if dy != 0:
            slope = dx / dy
        
        return slope


    def incrementSlopeCount(self, slope, slopeToCountMap, maxKey):
        if slopeToCountMap.ContainsKey(slope):
            slopeToCountMap[slope] += 1
        else:
            slopeToCountMap[slope] = 1
        if not slopeToCountMap.ContainsKey(maxKey) or  slopeToCountMap[slope] >= slopeToCountMap[maxKey]:
            maxKey = slope
        
        return maxKey


def set_Points(flds):
    p = [Point()]*len(flds)
    for i in range(len(flds)):
        tempStr = flds[i].split(",")
        p[i] = Point(int(tempStr[0]), int(tempStr[1]))
    return p


def output_Points(p):
    if len(p) == 0:
        return ""
    result = "[[" + str(p[0].x) + "," + str(p[0].y) + "]"
    for i in range(1,len(p)):
        result += ",[" + str(p[i].x) + "," + str(p[i].y) + "]"
    return result


def main():
    args = sys.argv
    argc = len(args)

    print("args[0] = %s %s" %(args[0], args[1]) )
    tempStr = args[1].rstrip()
    tempStr = tempStr.replace("[[","").replace("]]","")
    flds = tempStr.split("],[")
    p = set_Points(flds)
    print("p = %s" %(output_Points(p)))

    time0 = time.time()

    sl = Solution()
    print("common_node = %s" %(sl.maxPoints(p)))

    time1 = time.time()
    print("Execute time ... : %f[s]" %(time1 - time0))
    print()


if __name__ == "__main__":
    main()

File: Problems/0149_Max_Points_on_a_Line/Max_Points_on_a_Line.cs
using System;

// Definition for a point.
public class Point {
	public int x;
	public int y;
	public Point() { x = 0; y = 0; }
	public Point(int a, int b) { x = a; y = b; }
 }
 
public class Solution {
	public class Propet
	{
		public double slople;
		public int first_index_num;
		public bool slople_checked;

		public Propet()
		{
			slople = 0.0;
			first_index_num = -1;
			slople_checked = false;
		}
	};

    public int MaxPoints(Point[] points)
	{
		Propet[] pt = new Propet[points.Length];
		int i, j;
		for (i = 0; i < points.Length; i++) {
			pt[i] = new Propet();
			pt[i].slople = (double)points[i].y / points[i].x;		
		}

		for (i = 0; i < points.Length; i++) {
			if (pt[i].first_index_num != -1)
				continue;
			for (j = i + 1; j < points.Length; j++) {
				if (pt[j].first_index_num != -1)
					continue;
				if (pt[i].slople == pt[j].slople) {
					pt[i].first_index_num = pt[j].first_index_num = i;
				}
			}
		}

		for (i = 0; i < points.Length; i++) {
			Console.WriteLine("pt[" + i.ToString() + "] = " + pt[i].slople.ToString());
		}

		int count = 0, max = -1, max_index = -1;
		for (i = 0; i < points.Length; i++) {
			count = 0;
			for (j = 0; j < points.Length; j++) {
				if (pt[i].first_index_num == i) {
					count++;
				}
			}

			if (count > max) {
				max = count;
				max_index = i;
			}
		}

		int max_val = -1;
		for (i = 0; i < points.Length; i++) {
			if (pt[i].first_index_num == max_index) {
				if (points[i].y > max_val) {
					max_val = points[i].y;
				}
			}
		}

		return max_val;
    }

	private Point[] set_Points(string[] flds)
	{
		Point[] p = new Point[flds.Length];

		for (int i = 0; i < flds.Length; ++i) {
			string[] temp = flds[i].Split(",");
			p[i] = new Point(int.Parse(temp[0]), int.Parse(temp[1]));
		}

		return p;
	}

	private string output_Points(Point[] p)
	{
		if (p.Length == 0)
			return "";

		string result = "[[" + p[0].x + "," + p[0].y + "]";

		for (int i = 1; i < p.Length; i++) {
			result += ",[" + p[i].x + "," + p[i].y + "]";
		}

		result += "]";

		return result;
	}

	public void Main(string args)
	{
		string temp = args.Replace("[[","").Replace("]]","");
		string[] flds = temp.Split("],[");
		Point[] p = set_Points(flds);
		Console.WriteLine("p = " + output_Points(p));

		System.Diagnostics.Stopwatch sw = new System.Diagnostics.Stopwatch();
		sw.Start();

		Console.WriteLine("max = " + MaxPoints(p));
		
		sw.Stop();
		Console.WriteLine("Execute time ... " + sw.ElapsedMilliseconds.ToString() + "ms");
	}
}
